pkg/didcomm/crypto/jwe/authcrypt/encrypt.go

```diff
@@ -0,0 +1,368 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package authcrypt
+
+import (
+	"crypto"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+
+	josecipher "github.com/square/go-jose/v3/cipher"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/curve25519"
+	"golang.org/x/crypto/nacl/box"
+	errors "golang.org/x/xerrors"
+
+	"github.com/btcsuite/btcutil/base58"
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/poly1305"
+)
+
+// This package deals with Authcrypt encryption for Packing/Unpacking DID Comm exchange
+// Using Chacha20Poly1035 encryption/authentication
+
+// C20P Chacha20Poly1035 algorithm
+const C20P = "C20P"
+
+// XC20P XChacha20Poly1035 algorithm
+const XC20P = "XC20P"
+
+// RandReader is a cryptographically secure random number generator.
+var RandReader = rand.Reader
+
+type keyPair struct {
+	priv *[chacha20poly1305.KeySize]byte
+	pub  *[chacha20poly1305.KeySize]byte
+}
+
+// Crypter represents an Authcrypt Encrypter (Decrypter) that outputs/reads JWE envelopes
+type Crypter struct {
+	sender     keyPair
+	recipients []*[chacha20poly1305.KeySize]byte
+	alg        string
+	nonceSize  int
+}
+
+// Envelope represents a JWE envelope as per the Aries Encryption envelop specs
+type Envelope struct {
+	Protected  string      `json:"protected,omitempty"`
+	Recipients []Recipient `json:"recipients,omitempty"`
+	Aad        string      `json:"aad,omitempty"`
+	Iv         string      `json:"iv,omitempty"`
+	Tag        string      `json:"tag,omitempty"`
+	CipherText string      `json:"ciphertext,omitempty"`
+}
+
+// jweHeaders are the Protected JWE headers in a map format
+type jweHeaders map[string]string
+
+// Recipient is a recipient of an envelop including the shared encryption key
+type Recipient struct {
+	EncryptedKey string           `json:"encrypted_key,omitempty"`
+	Header       RecipientHeaders `json:"header,omitempty"`
+}
+
+// RecipientHeaders are the recipient headers
+type RecipientHeaders struct {
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Reply:
rename to RecipientHEADERS?

Reply:
No. I was referring to some of the fields.

Reply:
Also you should check Envelope fields.

Reply:
will update ..

Reply:
done..